corehq/apps/es/README.rst

```diff
@@ -0,0 +1,220 @@
+Elasticsearch App
+=================
+
+Adapter Design
+--------------
+
+The HQ Elastic adapter design came about due to the need for reindexing
+Elasticsearch indexes in a way that is transparent to parts of HQ that write to
+Elasticsearch (e.g. pillowtop).  Reindexing is necessary for making changes to
+index mappings, is a prerequisite to upgrading an Elasticsearch cluster, and is
+also needed for changing low-level index configurations (e.g. sharding).
+
+There is an existing procedure draft that documents the steps that were used on
+one occasion to reindex the ``case_search`` index.  This procedure leveraged a
+custom pillow to "backfill" the cloned index (i.e. initially populated using
+Elasticsearch Reindex API). That procedure only works for a subset of HQ
+Elasticsearch indexes, and is too risky to be considered as an ongoing Elastic
+maintenance strategy. There are several key constraints that an HQ reindexing
+procedure should meet which the existing procedure does not:
+
+- simple and robust
+- performed with standard maintenance practices
+- provides the ability to test and verify the integrity of a new index before it
+  is too late to be rejected
+- allows HQ Elasticsearch index state to remain decoupled from the
+  commcare-cloud codebase
+- is not disruptive -- does not prohibit any other kind of standard maintenance
+  that might come up while the operation is underway
+- is "fire and forget" -- does not require active polling of intermediate state
+  in order to progress the overall operation
+- is practical for third party HQ hosters to use
+
+One way to accomplish these constraints is to implement an "index multiplexing"
+feature in HQ, where Elasticsearch write operations are duplicated across two
+indexes. This design facilitates maintaining two up-to-date versions of any
+index (a primary read/write index and a secondary write-only index),
+allowing HQ to run in a "normal" state (i.e. not a custom "maintenance" state)
+while providing the ability to switch back and forth (swapping primary and
+secondary) before fully committing to abandoning one of them. Creating a copy of
+an index is the unavoidable nature of a reindex operation, and multiplexing
+allows safe switching from one to the other without causing disruptions or
+outages while keeping both up-to-date.
+
+The least disruptive way to accomplish a multiplexing design is with an adapter
```

Comment:
Can you speak more to how the adapter layer will allow us to do multiplexing. It's not clear from the current code who that will be accomplished.

Reply:
The adapter is a conduit through which all Elastic document actions go through (`index`, `update`, `delete`, `search`, etc). When the adapter is configured in "multiplexing mode", it will be able to translate every document action it receives into two actions: the "original" action is executed verbatim on the primary index, and an equivalent action is performed on the secondary index.  Because the adapter knows that _all_ actions for that index have come through it, it can thereby guarantee the integrity of documents that it maintains in the secondary index.

Reply:
> When the adapter is configured in "multiplexing mode"

This is the crux of my question. How will this be done? I had a discussion with @millerdev about this last week and he indicated that this would be achieved by creating a delegating adapter that would delegate to multiple other adapters. Is this how you intend to make multiplexing work?

Reply:
Yes.

Reply:
Would be great to add that into the doc.